storage/index_insert_queue.go

```diff
@@ -0,0 +1,258 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package storage
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3db/clock"
+	"github.com/m3db/m3ninx/doc"
+
+	"github.com/uber-go/tally"
+)
+
+var (
+	errIndexInsertQueueNotOpen             = errors.New("index insert queue is not open")
+	errIndexInsertQueueAlreadyOpenOrClosed = errors.New("index insert queue already open or is closed")
+	errNewSeriesIndexRateLimitExceeded     = errors.New("indexing new series eclipses rate limit")
+)
+
+type dbIndexInsertQueueState int
+
+const (
+	dbIndexInsertQueueStateNotOpen dbIndexInsertQueueState = iota
+	dbIndexInsertQueueStateOpen
+	dbIndexInsertQueueStateClosed
+)
+
+var (
+	defaultIndexBatchBackoff   = time.Second
+	defaultIndexPerSecondLimit = 10000
+)
+
+type dbIndexInsertQueue struct {
```

Comment:
Could you possibly use Genny to condense needing a copy?  After looking at a lot of stuff, actually doesn't seem like much is different between the two at all.

Reply:
Yeah they definitely share over 90% of code right now. I'm hesitant to make the change just now - I'd like to see how things are after we've got a fully functioning index. I'll open an issue to track for now, can come back and cleanup after.

Reply:
https://github.com/m3db/m3db/issues/517